network/peer-monitoring-service/client/src/lib.rs

```diff
@@ -151,6 +156,51 @@ async fn start_peer_monitor_with_state(
     }
 }
 
+/// Creates a new peer state for peers that don't yet have one
+fn create_states_for_new_peers(
+    node_config: &NodeConfig,
+    peer_monitor_state: &PeerMonitorState,
+    time_service: &TimeService,
+    connected_peers_and_metadata: &HashMap<PeerNetworkId, PeerMetadata>,
+) {
+    for peer_network_id in connected_peers_and_metadata.keys() {
+        let state_exists = peer_monitor_state
+            .peer_states
+            .read()
+            .contains_key(peer_network_id);
+        if !state_exists {
+            peer_monitor_state.peer_states.write().insert(
+                *peer_network_id,
+                PeerState::new(node_config.clone(), time_service.clone()),
+            );
+        }
+    }
+}
+
+/// Garbage collects peer states for peers that are no longer connected
+fn garbage_collect_peer_states(
+    peer_monitor_state: &PeerMonitorState,
+    connected_peers_and_metadata: &HashMap<PeerNetworkId, PeerMetadata>,
+) {
+    // Get the set of peers with existing states
+    let peers_with_existing_states: Vec<PeerNetworkId> = peer_monitor_state
+        .peer_states
+        .read()
+        .keys()
+        .cloned()
+        .collect();
+
+    // Remove the states for disconnected peers
+    for peer_network_id in peers_with_existing_states {
+        if !connected_peers_and_metadata.contains_key(&peer_network_id) {
+            peer_monitor_state
+                .peer_states
+                .write()
+                .remove(&peer_network_id);
+        }
+    }
+}
```

Comment:
Do you want to maybe, periodically garbage collect, rather than garbage collect on every connection?

Perhaps on every X connections or after X period of time.  the `sample!()` macro can be useful for this.

Reply:
Yeah, this is currently only called by the peer monitor (every peer monitor loop duration -- which I think defaults to 1 second), so it's not happening super frequently today 😄